wakuv2/nwaku.go

```diff
@@ -2377,547 +2329,737 @@ func wakuNew(nodeKey *ecdsa.PrivateKey,
+		topicHealthStatusChan:           make(chan peermanager.TopicHealthStatus, 100),
+		connectionNotifChan:             make(chan node.PeerConnection, 20),
+		connStatusSubscriptions:         make(map[string]*types.ConnStatusSubscription),
+		ctx:                             ctx,
+		cancel:                          cancel,
+		wg:                              sync.WaitGroup{},
+		dnsAddressCache:                 make(map[string][]dnsdisc.DiscoveredNode),
+		dnsAddressCacheLock:             &sync.RWMutex{},
+		dnsDiscAsyncRetrievedSignal:     make(chan struct{}),
+		storeMsgIDs:                     make(map[gethcommon.Hash]bool),
+		timesource:                      ts,
+		storeMsgIDsMu:                   sync.RWMutex{},
+		logger:                          logger,
+		discV5BootstrapNodes:            nwakuCfg.Discv5BootstrapNodes,
+		onHistoricMessagesRequestFailed: onHistoricMessagesRequestFailed,
+		onPeerStats:                     onPeerStats,
+		onlineChecker:                   onlinechecker.NewDefaultOnlineChecker(false).(*onlinechecker.DefaultOnlineChecker),
+		sendQueue:                       publish.NewMessageQueue(1000, cfg.UseThrottledPublish),
+	}, nil
 
-	errMsg := "error wakuNew: " + C.GoStringN(C.getMyCharPtr(resp), C.int(C.getMyCharLen(resp)))
-	return nil, errors.New(errMsg)
 }
 
-func (self *Waku) WakuStart() error {
-
-	var resp = C.allocResp()
-	defer C.freeResp(resp)
-	C.cGoWakuStart(self.wakuCtx, resp)
-
-	if C.getRet(resp) == C.RET_OK {
-		return nil
-	}
-	errMsg := "error WakuStart: " + C.GoStringN(C.getMyCharPtr(resp), C.int(C.getMyCharLen(resp)))
-	return errors.New(errMsg)
+//export globalEventCallback
+func globalEventCallback(callerRet C.int, msg *C.char, len C.size_t, userData unsafe.Pointer) {
+	// This is shared among all Golang instances
+	// TODO-nwaku
+	// self := Waku{wakuCtx: userData}
+	// self.MyEventCallback(callerRet, msg, len)
 }
 
-func (self *Waku) WakuStop() error {
-	var resp = C.allocResp()
-	defer C.freeResp(resp)
-	C.cGoWakuStop(self.wakuCtx, resp)
-
-	if C.getRet(resp) == C.RET_OK {
-		return nil
-	}
-	errMsg := "error WakuStop: " + C.GoStringN(C.getMyCharPtr(resp), C.int(C.getMyCharLen(resp)))
-	return errors.New(errMsg)
+func (self *Waku) MyEventCallback(callerRet C.int, msg *C.char, len C.size_t) {
+	fmt.Println("Event received:", C.GoStringN(msg, C.int(len)))
 }
-func (self *Waku) WakuDestroy() error {
-	var resp = C.allocResp()
-	defer C.freeResp(resp)
-	C.cGoWakuDestroy(self.wakuCtx, resp)
 
-	if C.getRet(resp) == C.RET_OK {
-		return nil
-	}
-	errMsg := "error WakuDestroy: " + C.GoStringN(C.getMyCharPtr(resp), C.int(C.getMyCharLen(resp)))
-	return errors.New(errMsg)
+type response struct {
+	err   error
+	value any
 }
 
-func (self *Waku) StartDiscV5() error {
-	var resp = C.allocResp()
-	defer C.freeResp(resp)
-	C.cGoWakuStartDiscV5(self.wakuCtx, resp)
+type WakuNode struct {
+	wakuCtx unsafe.Pointer
+	cancel  context.CancelFunc
+	newCh   chan *request
+}
```

Comment:
Something to consider in upcoming PRs.
Strictly speaking, the `wakuCtx` contains a WakuNode; discovery; and other items.

```suggestion
type Waku struct {
	wakuCtx unsafe.Pointer
	cancel  context.CancelFunc
	newCh   chan *request
}
```
---------

We will also need to remove the current `type Waku struct`. See: https://github.com/status-im/status-go/blob/6c763564154eb3179bdaca34f7340cfe19d32f99/wakuv2/nwaku.go#L435